### DISCUSSION TITLE
Update `cache-first` query after mutation

### DISCUSSION THREAD

**Discussion by beeirl** [urql-graphql/urql#3188] [Q&A]
I am working on a React SPA. I have a `OrganizationView` with the following query + mutation:

```
 const OrganizationQuery = gql`
	query ($id: ID!) {
		organization(id: $id) {
			id
			name
		}
	}
`
const UpdateOrganizationMutation = gql`
	mutation ($id: ID!, $name: String!) {
		updateOrganization(id: $id, name: $name) {
			id
			name
		}
	}
`

const [organizationQuery] = useQuery(OrganizationQuery)
const [_, updateOrganization] = useMutation(UpdateOrganizationMutation)
```

The `organizationQuery` uses `cache-first` as the default `requestPolicy`.
If I call the `updateOrganization` mutation after the initial fetch of the `organizationQuery` while the `OrganizationView` is mounted everything works as expected. The `updateOrganization` mutation would update the cache automatically without the need of manually calling an `update` function to update the cache accordingly.
However, if I let's say navigate to a different view and navigate back to the `OrganizationView`, the `organizationQuery` would fetch its data from the cache (since it's `cache-first`). So far so good. If I now call the `updateOrganization` mutation again, the cache won't update and the `organizationQuery` data becomes outdated. 
If I'd change the `requestPolicy` of the `organizationQuery` to `cache-and-network` it just works fine. Not a big fan of doing this since this will lead to unnecessary network requests.
Is this the expected behaviour or am I missing something here?

**Comment by kitten** [comment]
What you’re expecting to happen is definitely expected behaviour.

Assuming you’re using the document cache (although the same is expected on Graphcache; the normalised cache) this is the relevant code for this: https://github.com/urql-graphql/urql/blob/cb628ca4e9207b1928a5e137fddbd14c33afb6e4/packages/core/src/exchanges/cache.ts#L132-L147

i.e. what happens is the typenames of the mutations will invalidate matching queries regardless of whether they're currently mounted or not, which means, on the next mount the query _should_ not be cached.

It's hard for me however to explain why this isn't happening, but it may be useful to you to use the urql devtools (or the debug exchange) and check the order of operations, since you should see exactly whether the query’s cached result is invalidated or not 🤔

**Reply by beeirl** [reply]
@kitten thank you for your explanation. I've indeed observed the behaviour with the devtools and the query's cached result is never invalidated upon re-mount with `cache-first`. Again, works completely fine with `cache-and-network`. Want me to put together a reproducible example?

**Reply by kitten** [reply]
I can take a look if you can reproduce it in a smaller example, yea ✌️

**Reply by beeirl** [reply]
Alright heres the [CSB](https://codesandbox.io/s/strange-chaplygin-mssecx). It seems to be caused by `@urql/exchange-graphcache@6.0.2`. Problem doesn't exist on `6.0.1`. Will check what has changed in a second.

**Reply by kitten** [reply]
Oh wow, yep, that's not great. I reckon a refactor has gone awry and our unit tests didn't catch that. I can only promise a fix tomorrow, but that's pretty high priority 😅

**Reply by beeirl** [reply]
All good, glad I was able to bring this to your attention at least 😅

**Reply by kitten** [reply]
This was pretty easy to spot and a regression hidden away behind teardowns, so a fix should go out with this PR :v: https://github.com/urql-graphql/urql/pull/3193

**Comment by kitten** [comment]
This should be resolved and published as part of `@urql/exchange-graphcache@6.0.4`

**Reply by beeirl** [reply]
Appreciate it!
